File: ios_app/iphone/Fs.swift
import SwiftUI

let Fs__TITLE_FONT_SIZE = 27.0 // Golden ratio to lists text
let Fs__TITLE_FONT_WEIGHT: Font.Weight = .heavy
let Fs__BUTTON_FONT_SIZE = 16.0

class Fs: ObservableObject {

    @Published var items = [Fs__Item<AnyView>]()

    func show<Content: View>(
        @ViewBuilder content: @escaping (Binding<Bool>) -> Content
    ) {
        items.append(
            Fs__Item(
                content: { isPresented in
                    AnyView(content(isPresented))
                }
            )
        )
    }
}

extension View {

    func attachFs() -> some View {
        modifier(Fs__Modifier())
    }
}

private let fsAnimation = Animation.easeInOut(duration: 0.20)

///
///

struct Fs__Item<Content: View>: View, Identifiable {

    @ViewBuilder var content: (Binding<Bool>) -> Content

    @EnvironmentObject private var fs: Fs
    @State private var isPresented = false

    let id = UUID().uuidString

    var body: some View {

        ZStack {

            if isPresented {

                ZStack {

                    VStack {
                        HStack {
                            Spacer()
                        }
                        Spacer()
                    }
                    .background(c.bg)
                    .ignoresSafeArea()

                    content($isPresented)
                        .ignoresSafeArea(.container) // Keep keyboard's paddings
                        .onDisappear {
                            fs.items.removeAll {
                                $0.id == id
                            }
                        }
                }
                .transition(.opacity)
            }
        }
        .animation(fsAnimation, value: isPresented)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
        .onAppear {
            isPresented = true
            hideKeyboard()
        }
    }
}

private struct Fs__Modifier: ViewModifier {

    @StateObject private var fs = Fs()

    func body(content: Content) -> some View {
        ZStack {
            content
            ForEach(fs.items) { wrapper in
                wrapper
            }
        }
        .environmentObject(fs)
    }
}

///

struct Fs__CloseButton: View {

    let onClick: () -> Void

    var body: some View {

        Button(
            action: {
                onClick()
            },
            label: {
                Image(systemName: "xmark")
                    .font(.system(size: 14, weight: .medium))
                    .foregroundColor(c.tertiaryText)
            }
        )
        .frame(width: 32, height: 32)
        .background(roundedShape.fill(c.fg))
    }
}

struct Fs__Header<Content: View>: View {

    let scrollToHeader: Int
    @ViewBuilder var content: () -> Content

    var body: some View {

        // .bottom for divider
        ZStack(alignment: .bottom) {

            content()

            DividerBgScroll(scrollToHeader: scrollToHeader)
        }
        .safeAreaPadding(.top)
    }
}

struct Fs__HeaderTitle: View {

    let title: String
    let scrollToHeader: Int
    let onClose: () -> Void

    var body: some View {

        Fs__Header(
            scrollToHeader: scrollToHeader
        ) {

            HStack {

                HeaderTitle(
                    title: title
                )

                Spacer()

                Fs__CloseButton(
                    onClick: {
                        onClose()
                    }
                )
                .padding(.trailing, H_PADDING)
            }
            .padding(.bottom, 6)
        }
    }
}

struct Fs__HeaderAction: View {

    let title: String
    let actionText: String
    let scrollToHeader: Int
    let onCancel: () -> Void
    let onDone: () -> Void

    private var bgAlpha: Double {
        (Double(scrollToHeader) / 30).limitMinMax(0, 1)
    }

    var body: some View {

        Fs__Header(
            scrollToHeader: scrollToHeader
        ) {

            VStack(alignment: .leading) {

                Button(
                    action: {
                        onCancel()
                    },
                    label: {
                        Text("Cancel")
                            .foregroundColor(c.textSecondary)
                            .font(.system(size: 15, weight: .light))
                            .padding(.leading, H_PADDING)
                    }
                )
                .padding(.bottom, 2)

                HStack {

                    HeaderTitle(
                        title: title
                    )

                    Spacer()

                    Button(
                        action: {
                            onDone()
                        },
                        label: {
                            Text(actionText)
                                .foregroundColor(c.text)
                                .font(.system(size: Fs__BUTTON_FONT_SIZE, weight: Fs__TITLE_FONT_WEIGHT))
                                .padding(.horizontal, 10)
                                .padding(.vertical, 6)
                                .background(roundedShape.fill(c.blue))
                        }
                    )
                    .padding(.trailing, H_PADDING)
                }
                .padding(.bottom, 6)
            }
            .padding(.leading, halfDpCeil)
        }
    }
}

struct Fs__HeaderClose: View {

    let onClose: () -> Void

    var body: some View {

        HStack {

            Spacer()

            Fs__CloseButton {
                onClose()
            }
            .padding(.trailing, H_PADDING)
        }
        .safeAreaPadding(.top)
        .padding(.top, H_PADDING)
        .zIndex(2)
    }
}

private struct HeaderTitle: View {

    let title: String

    var body: some View {
        Text(title)
            .foregroundColor(c.text)
            .font(.system(size: Fs__TITLE_FONT_SIZE, weight: Fs__TITLE_FONT_WEIGHT))
            .padding(.leading, H_PADDING)
    }
}

File: android_app/src/main/java/me/timeto/app/ui/Fs.kt
package me.timeto.app.ui

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.ScrollableState
import androidx.compose.foundation.layout.*
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.zIndex
import me.timeto.app.*
import me.timeto.app.R

private val enterAnimation = fadeIn(spring(stiffness = Spring.StiffnessMedium))
private val exitAnimation = fadeOut(spring(stiffness = Spring.StiffnessMedium))

val Fs__TITLE_FONT_SIZE = 26.sp // Golden ratio to lists text
val Fs__TITLE_FONT_WEIGHT = FontWeight.ExtraBold
val Fs__BUTTON_FONT_SIZE = 15.sp

object Fs {

    fun show(
        content: @Composable (WrapperView.Layer) -> Unit,
    ) {

        WrapperView.Layer(
            enterAnimation = enterAnimation,
            exitAnimation = exitAnimation,
            alignment = Alignment.BottomCenter,
            onClose = {},
            content = { layer ->
                Box(
                    modifier = Modifier
                        .pointerInput(Unit) {}
                ) {
                    content(layer)
                }
            }
        ).show()
    }
}

@Composable
fun Fs__CloseButton(
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    ZStack(
        modifier = modifier
            .size(31.dp)
            .clip(roundedShape)
            .background(c.fg)
            .clickable {
                onClick()
            },
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            painterResource(R.drawable.sf_xmark_small_medium),
            contentDescription = "Close",
            tint = c.tertiaryText,
            modifier = Modifier
                .size(11.dp),
        )
    }
}

@Composable
fun Fs__Header(
    scrollState: ScrollableState?,
    content: @Composable () -> Unit,
) {

    ZStack(
        modifier = Modifier
            .padding(top = (LocalContext.current as MainActivity).statusBarHeightDp),
    ) {

        content()

        DividerBgScroll(
            scrollState = scrollState,
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(horizontal = H_PADDING),
        )
    }
}

@Composable
fun Fs__HeaderTitle(
    title: String,
    scrollState: ScrollableState?,
    onClose: () -> Unit,
) {

    Fs__Header(
        scrollState = scrollState,
    ) {

        HStack(
            modifier = Modifier
                .padding(top = 20.dp, bottom = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {

            HeaderTitle(
                title = title,
            )

            Fs__CloseButton(
                modifier = Modifier
                    .padding(end = H_PADDING),
            ) {
                onClose()
            }
        }
    }
}

@Composable
fun Fs__HeaderAction(
    title: String,
    actionText: String,
    scrollState: ScrollableState?,
    onCancel: () -> Unit,
    onDone: () -> Unit,
) {

    Fs__Header(
        scrollState = scrollState,
    ) {

        VStack(
            modifier = Modifier
                .padding(start = halfDpCeil),
        ) {

            Text(
                text = "Cancel",
                modifier = Modifier
                    .offset(y = 1.dp)
                    .padding(start = H_PADDING_HALF, top = 12.dp)
                    .clip(roundedShape)
                    .clickable { onCancel() }
                    .padding(horizontal = H_PADDING_HALF),
                color = c.textSecondary,
                fontWeight = FontWeight.Light,
                fontSize = 14.sp,
            )

            HStack(
                modifier = Modifier
                    .padding(bottom = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {

                HeaderTitle(
                    title = title,
                )

                Text(
                    text = actionText,
                    modifier = Modifier
                        .padding(end = H_PADDING)
                        .clip(roundedShape)
                        .background(c.blue)
                        .clickable {
                            onDone()
                        }
                        .padding(
                            horizontal = 10.dp,
                            vertical = 3.dp,
                        ),
                    color = c.text,
                    fontSize = Fs__BUTTON_FONT_SIZE,
                    fontWeight = Fs__TITLE_FONT_WEIGHT,
                )
            }
        }
    }
}

@Composable
fun Fs__HeaderClose(
    onClose: () -> Unit,
) {

    HStack(
        modifier = Modifier
            .padding(top = (LocalContext.current as MainActivity).statusBarHeightDp + H_PADDING)
            .zIndex(1f),
    ) {

        SpacerW1()

        Fs__CloseButton(
            modifier = Modifier
                .padding(end = H_PADDING),
        ) {
            onClose()
        }
    }
}

@Composable
private fun RowScope.HeaderTitle(
    title: String,
) {
    Text(
        text = title,
        modifier = Modifier
            .padding(start = H_PADDING)
            .weight(1f),
        fontSize = Fs__TITLE_FONT_SIZE,
        fontWeight = Fs__TITLE_FONT_WEIGHT,
        color = c.text,
    )
}

//
// Bottom

@Composable
fun Fs__BottomBar(
    content: @Composable () -> Unit,
) {

    VStack(
        modifier = Modifier
            .navigationBarsPadding(),
    ) {

        DividerBg()

        content()
    }
}
